### PR TITLE
fix(gossipsub): Attempt to publish to at least mesh_n peers

### PR DESCRIPTION
**Pull request by jxs**
## Description

With flood published disabled we've noticed that it can be the case that we have connected peers on topics but these peers are not in our mesh (perhaps due to their own mesh requirements). Currently, we fail to publish the message if there are no peers in our mesh.

This PR adjusts this logic to always attempt to publish to at least mesh_n peers. If we have peers that are subscribed to a topic, we will now attempt to publish messages to them (provided they have the required score).

This PR also simplies the peer and respective topics by moving the topic list each peer has subscribed to `PeerConnections` and removing both `peer_topics` and `topic_peers` from the main `Behaviour`.
Per commit review is suggested.

### CODE REVIEW

File: protocols/gossipsub/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
 ## 0.47.0
 
+- Attempt to publish to at least mesh_n peers when flood publish is disabled.
+  See [PR XXXX](https://github.com/libp2p/rust-libp2p/pull/XXXX).
```
Comment (dariusc93) [inline protocols/gossipsub/CHANGELOG.md]:
```suggestion
  See [PR 5578](https://github.com/libp2p/rust-libp2p/pull/5578).
```

File: protocols/gossipsub/src/behaviour.rs [hunk 2]
```diff
@@ -1857,29 +1863,20 @@ where
...
             match subscription.action {
                 SubscriptionAction::Subscribe => {
-                    if peer_list.insert(*propagation_source) {
+                    if peer.topics.insert(topic_hash.clone()) {
                         tracing::debug!(
                             peer=%propagation_source,
                             topic=%topic_hash,
                             "SUBSCRIPTION: Adding gossip peer to topic"
                         );
                     }
```
Comment (ackintosh) [inline protocols/gossipsub/src/behaviour.rs]: Should we add `m.inc_topic_peers(topic_hash);` here?

```suggestion
                    if peer.topics.insert(topic_hash.clone()) {
                        tracing::debug!(
                            peer=%propagation_source,
                            topic=%topic_hash,
                            "SUBSCRIPTION: Adding gossip peer to topic"
                        );
                    }

                    if let Some(m) = self.metrics.as_mut() {
                        m.inc_topic_peers(topic_hash);
                    }
```

**Review by ackintosh — COMMENTED** [review]
Looks good to me! I left a small comment. 🙂

File: protocols/gossipsub/src/behaviour.rs (hunk 2 above)
Comment (jxs) [reply protocols/gossipsub/src/behaviour.rs]: yup makes sense, also updated the other place where we inc the metrics to only do it when we actually insert in `self.topics`.
Thanks Akihito!

**Review by ackintosh — APPROVED** [review]

File: protocols/gossipsub/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 ## 0.47.0
```
Comment (dariusc93) [inline protocols/gossipsub/CHANGELOG.md]:
```suggestion
## 0.47.1
```

**Pull request merged**